lib/content-negotiation.js

```diff
@@ -25,10 +26,18 @@ const SUPPORTED_MEDIA_TYPES = [
     "application/x-www-form-urlencoded",
     "application/cloudevents+json",
 ]; // In preference order
-const textDecoder = new TextDecoder("utf8");
+
+const utf8TextDecoder = new TextDecoder("UTF-8");
 const textEncoder = new TextEncoder();
 
-const jsonUnmarshall = (buffer) => JSON.parse(textDecoder.decode(buffer));
+const jsonUnmarshall = (buffer, charset) => {
+    if (charset && !charset.toLocaleLowerCase("en-US").startsWith("utf-")) {
+        throw new Error(
+            `JSON only accepts Unicode charsets, given: ${charset}`
+        );
+    }
+    return JSON.parse(resolveDecoder(charset, () => "utf8").decode(buffer));
```

Comment:
unless we know we can handle utf-16 encoded json, let's require that it be utf-8

```suggestion
    if (charset && charset.toLocaleLowerCase("en-US") !== "utf-8") {
        throw new Error(
            `JSON only accepts the UTF-8 charset, given: ${charset}`
        );
    }
    return JSON.parse(utf8TextDecoder.decode(buffer));
```

Reply:
I assume JSON.parse/stringify properly handle these, why the restriction? I could actually some tests with UTF-16 / UTF-32 JSON so we can rest assured.

Reply:
UTF-32 [is not supported](https://github.com/nodejs/node/issues/5956) despite JSON [supporting it](https://tools.ietf.org/html/rfc7159#section-8.1).
I added tests for UTF-16(LE) and UTF-16BE.
I rewrote the validation logic so that only the supported encodings are accepted.